### PR TITLE
Detect builder repo and ref from oidc token

### PR DESCRIPTION
**Pull request by asraa**
Signed-off-by: Asra Ali <asraa@google.com>

Fixes https://github.com/slsa-framework/slsa-github-generator-go/issues/24

### CODE REVIEW

**Comment by asraa** [comment]
Ready! @laurentsimon 
Works at https://github.com/asraa/slsa-on-github-test/actions/runs/2072678668
much easier to just update the ref once!

File: .github/workflows/builder.yml [hunk 1]
```diff
@@ -54,18 +51,37 @@ jobs:
...
+      contents: read
+    outputs:
+      builder_repo: ${{ steps.detect.outputs.builder_repo }}
+      builder_ref: ${{ steps.detect.outputs.builder_ref }}
+    steps:
+      - name: Detect the builder ref
+        id: detect
+        shell: bash
+        run: | 
+          export OIDC_TOKEN=$(curl -sS -H "Authorization: Bearer $ACTIONS_ID_TOKEN_REQUEST_TOKEN" "$ACTIONS_ID_TOKEN_REQUEST_URL" | jq -r '.value')
```
Comment (laurentsimon) [inline .github/workflows/builder.yml:67]: can you add a specific audience, like in https://github.com/slsa-framework/slsa-github-generator-go/blob/main/pkg/provenance.go#L291? Maybe you can also update the one we use in this file to something more specific, e.g., `slsa-framework/slsa-github-generator-go/_some_usecase`. I want to be sure a leak of this token does not allow forging provenance signatures.

Comment (laurentsimon) [inline .github/workflows/builder.yml:67]: although it's unlikely to fail in practice, can we add some error handling, something like https://github.com/ossf/scorecard-action/blob/36a22b695be1a46ca0e2bc47549cd4d311f6c0ee/entrypoint.sh#L50-L55?

File: .github/workflows/builder.yml [hunk 2]
```diff
@@ -54,18 +51,37 @@ jobs:
...
+          export WORKFLOW_REF=$(echo $OIDC_TOKEN | cut -d "." -f2 | base64 -d | jq -r '.job_workflow_ref')
+          echo "::set-output name=builder_repo::$(echo $WORKFLOW_REF | cut -d "@" -f1 | cut -d '/' -f1-2)"
```
Comment (laurentsimon) [inline .github/workflows/builder.yml:69]: Can we add a test on whether the `$WORKFLOW_REF` is empty?

**Review by laurentsimon — APPROVED** [review]
Thanks!

File: .github/workflows/builder.yml (hunk 1 above)
Comment (asraa) [reply .github/workflows/builder.yml:67]: done!

Comment (asraa) [reply .github/workflows/builder.yml]: I made it the fully qualified URL here ...
https://github.com/slsa-framework/slsa-github-generator-go

Comment (laurentsimon) [reply .github/workflows/builder.yml]: even better :-)

**Comment by asraa** [comment]
FYI doesn't want to issue for "Failed to get OIDC token from GitHub, response 403: Can't issue ID_TOKEN for audience 'https://github.com/slsa-framework/slsa-github-generator-go'."

might be because of the full qualification or a bad char... seemed like i had seen azure examples for this, will try to find docs about it

**Comment by asraa** [comment]
all good now -- i can't get fully qualified URI, and I suspect the 403 is because the context of the caller is not for the reusable workflow github repo :/

**Comment by laurentsimon** [comment]
> all good now -- i can't get fully qualified URI, and I suspect the 403 is because the context of the caller is not for the reusable workflow github repo :/

Gotcha, that's probably good enough.
